src/apprt/gtk/headerbar.zig

```diff
@@ -4,93 +4,64 @@ const c = @import("c.zig").c;
 const Window = @import("Window.zig");
 const adwaita = @import("adwaita.zig");
 
-const AdwHeaderBar = if (adwaita.versionAtLeast(0, 0, 0)) c.AdwHeaderBar else void;
+const HeaderBarAdw = @import("headerbar_adw.zig").HeaderBarAdw;
+const HeaderBarGtk = @import("headerbar_gtk.zig").HeaderBarGtk;
 
 pub const HeaderBar = union(enum) {
-    adw: *AdwHeaderBar,
-    gtk: *c.GtkHeaderBar,
+    adw: HeaderBarAdw,
+    gtk: HeaderBarGtk,
 
-    pub fn init(window: *Window) HeaderBar {
-        if ((comptime adwaita.versionAtLeast(1, 4, 0)) and
-            adwaita.enabled(&window.app.config))
-        {
-            return initAdw(window);
+    pub fn init(self: *HeaderBar) void {
+        const window: *Window = @fieldParentPtr("headerbar", self);
+        if ((comptime adwaita.versionAtLeast(1, 4, 0)) and adwaita.enabled(&window.app.config)) {
+            HeaderBarAdw.init(self);
+        } else {
+            HeaderBarGtk.init(self);
         }
 
-        return initGtk();
-    }
-
-    fn initAdw(window: *Window) HeaderBar {
-        const headerbar = c.adw_header_bar_new();
-        c.adw_header_bar_set_title_widget(@ptrCast(headerbar), @ptrCast(c.adw_window_title_new(c.gtk_window_get_title(window.window) orelse "Ghostty", null)));
-        return .{ .adw = @ptrCast(headerbar) };
-    }
-
-    fn initGtk() HeaderBar {
-        const headerbar = c.gtk_header_bar_new();
-        return .{ .gtk = @ptrCast(headerbar) };
+        if (!window.app.config.@"gtk-titlebar" or !window.app.config.@"window-decoration")
+            self.setVisible(false);
     }
 
     pub fn setVisible(self: HeaderBar, visible: bool) void {
-        c.gtk_widget_set_visible(self.asWidget(), @intFromBool(visible));
+        switch (self) {
+            .adw => |adw| adw.setVisible(visible),
+            .gtk => |gtk| gtk.setVisible(visible),
```

Comment:
Could use the `inline else` pattern in order to reduce typing

Reply:
Fixed.